### PR TITLE
Apply implicit indents to `WHEN` blocks and introduce `indented_then_contents`

### LINKED ISSUE #4754: Allow for more compact representation of multi-line `CASE .. WHEN` statements
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Complicated `CASE .. WHEN` statements are currently formatted with plenty of whitespace. For more complex statements, this is, however, undesirable. Eventually, there are two features that should enable that:

1. Allow implicit indents for `WHEN` blocks, i.e.

   ```sql
   SELECT
       col1,
       CASE
           WHEN col2 = 1
           THEN col2 + 1
       END AS col2
   FROM table1
   ```

   should be formattable as follows:

   ```sql
   SELECT
       col1,
       CASE WHEN col2 = 1
           THEN col2 + 1
       END AS col2
   FROM table1
   ```

   This should require setting `indented_then = false` along with `allow_implicit_indents = true`.

2. Allow nested `CASE .. WHEN` statements to use less whitespace, i.e.

   ```sql
   SELECT
       col1,
       CASE
           WHEN col2 = 1
           THEN CASE
                   WHEN col2 = 2
                   THEN col2 + 1
               END
       END AS col2
   FROM table1
   ```

   should be formattable as follows:

   ```sql
   SELECT
       col1,
       CASE
           WHEN col2 = 1
           THEN CASE
               WHEN col2 = 2
               THEN col2 + 1
           END
       END AS col2
   FROM table1
   ```

   This should require setting `indented_then = false` along with a new option `indented_then_contents = false`.

### Use case

_No response_

### Dialect

n/a

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by borchero**
### Brief summary of the change made
Resolves #4754.

### Are there any other side effects of this change that we should be aware of?

No impact on existing code.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by borchero** [comment]
FYI @svengiegerich

File: test/fixtures/rules/std_rule_cases/LT02-indent.yml [hunk 1]
```diff
@@ -1941,3 +1941,32 @@ test_fail_issue_4680:
         {% else %}
         col1 > 0
       {% endif %}
+
```
Comment (WittierDinosaur) [inline test/fixtures/rules/std_rule_cases/LT02-indent.yml:1944]: can you add a test with a fail_str?

Comment (borchero) [reply test/fixtures/rules/std_rule_cases/LT02-indent.yml:1944]: @WittierDinosaur I replaced the `pass_str` with `fix_str` and added a `fail_str` with the unwanted indentation

File: src/sqlfluff/core/default_config.cfg [hunk 2]
```diff
@@ -64,6 +64,7 @@ indented_ctes = False
 indented_using_on = True
 indented_on_contents = True
 indented_then = True
+indented_then_contents = True
```
Comment (WittierDinosaur) [inline src/sqlfluff/core/default_config.cfg:67]: Let's not include this in the default config for now

Comment (borchero) [reply src/sqlfluff/core/default_config.cfg:67]: Why not? This just recovers the current behavior. If I'm not mistaken, the formatting for existing queries is changed if this is not included?

**Comment by borchero** [comment]
@WittierDinosaur can you have another look at this? This PR blocks our company from adopting SQLFluff v2.

**Comment by borchero** [comment]
@WittierDinosaur friendly ping 🐧 possibly also @alanmcruickshank

File: src/sqlfluff/core/default_config.cfg (hunk 2 above)
Comment (WittierDinosaur) [reply src/sqlfluff/core/default_config.cfg:67]: Oh yeah - you're right!

**Review by WittierDinosaur — APPROVED** [review]
LGTM

**Pull request merged**